internal/controller/promotions/promotions.go

```diff
@@ -326,16 +328,26 @@ func (r *reconciler) serializedSync(
 
 				phase := kargoapi.PromotionPhaseSucceeded
 				phaseError := ""
-				if err = r.promoteFn(
-					promoCtx,
-					promo.Spec.Stage,
-					promo.Namespace,
-					promo.Spec.Freight,
-				); err != nil {
-					phase = kargoapi.PromotionPhaseErrored
-					phaseError = err.Error()
-					logger.Errorf("error executing Promotion: %s", err)
-				}
+
+				func() {
+					defer func() {
+						if err := recover(); err != nil {
+							logger.Errorf("Promotion panic: %v", err)
+							phase = kargoapi.PromotionPhaseErrored
+							phaseError = fmt.Sprintf("%v", err)
+						}
+					}()
+					if err = r.promoteFn(
+						promoCtx,
+						promo.Spec.Stage,
+						promo.Namespace,
+						promo.Spec.Freight,
+					); err != nil {
+						phase = kargoapi.PromotionPhaseErrored
+						phaseError = err.Error()
+						logger.Errorf("error executing Promotion: %s", err)
+					}
+				}()
```

Comment:
Are these changes still required in light of the recover panic option?

Reply:
Yes I intentionally left this in, because this has the added benefit of patching the promotion so that promo enters a Error state. Without this, the Promos that cause the panic, will continue to reconcile in a loop until the crash is dealt with. Unlike Stages/Applications (which do have to reconcile forever), at least with Promos we can stop the bleeding.

Reply:
That makes sense!